frontend/components/UpgradeURD.tsx

```diff
@@ -0,0 +1,315 @@
+  },
+  {
+    title: 'Migrate your GRAVE to the new forwarder',
+    completeText: { text: '', address: null },
+    complete: false,
+  },
+];
+
+export const UpgradeURD = ({
+  oldForwarderAddress,
+}: {
+  oldForwarderAddress: string;
+}) => {
+  const walletContext = useContext(WalletContext);
+  const { account, networkConfig, setURDLsp7, setURDLsp8, mainUPController } =
+    walletContext;
+  const [isSubmitting, setIsSubmitting] = useState<boolean>(false);
+
+  const bgColor = useColorModeValue('light.green.brand', 'dark.purple.200');
+  const provider = new ethers.providers.Web3Provider(window.lukso);
+  const graveForwarder = new ethers.Contract(
+    networkConfig.universalGraveForwarder,
+    LSP1GraveForwarderAbi.abi,
+    provider
+  ) as LSP1GraveForwarder;
+
+  const oldForwarder = new ethers.Contract(
+    oldForwarderAddress,
+    LSP1GraveForwarderAbi.abi,
+    provider
+  ) as LSP1GraveForwarder;
+
+  const signer = provider.getSigner();
+
+  const [leaveSteps, setLeaveSteps] = React.useState([...initialLeavingSteps]);
+  const [leavingStep, setLeavingStep] = useState<number>(-1);
+
+  const { activeStep: activeLeavingStep, setActiveStep: setActiveLeavingStep } =
+    useSteps({
+      index: -1,
+      count: leaveSteps.length,
+    });
+
+  const toast = useToast();
+
+  const handleUpgrade = async () => {
+    setIsSubmitting(true);
+    setLeavingStep(0);
+
+    try {
+      await updateBECPermissions(account!, mainUPController!, provider, signer);
+    } catch (e: any) {
+      console.error('Error updating permissions', e);
+      toast({
+        title: `Error updating permissions: ${e.message}`,
+        status: 'error',
+        position: 'bottom-left',
+        duration: 9000,
+        isClosable: true,
+      });
+      setLeavingStep(0);
+      setIsSubmitting(false);
+      return;
+    }
+
+    setLeavingStep(1);
+
+    try {
+      await resetLSPDelegates(provider, signer, oldForwarderAddress);
+    } catch (e: any) {
+      console.error('Error resetting LSP delegates', e);
+      toast({
+        title: `Error resetting LSP delegates: ${e.message}`,
+        status: 'error',
+        position: 'bottom-left',
+        duration: 9000,
+        isClosable: true,
+      });
+      setLeavingStep(0);
+      setIsSubmitting(false);
+      return;
+    }
+
+    setLeavingStep(2);
+
+    try {
+      await setForwarderAsLSPDelegate(
+        account!,
+        networkConfig.universalGraveForwarder,
+        signer,
+        provider
+      );
+    } catch (e: any) {
+      console.error('Error setting forwarder as LSP delegate', e);
+      toast({
+        title: `Error setting forwarder as LSP delegate ${e.message}`,
+        status: 'error',
+        position: 'bottom-left',
+        duration: 9000,
+        isClosable: true,
+      });
+      setLeavingStep(0);
+      setIsSubmitting(false);
+    }
+
+    setLeavingStep(3);
+
+    try {
+      const existingGrave = await oldForwarder.connect(signer).getGrave();
```

Comment:
given that you've set up the rest of the step as imports from urdUtils, this should be similarly abstracted in such a way. @democrazyfi

Reply:
I thought we'd do that just for things that need to be shared but perhaps for consistencey we can do that